direct/nn/conjgradnet/conjgrad.py

```diff
@@ -0,0 +1,323 @@
+                bk = _PRP(rk_new, rk_old, dim)
+            elif self.bk_update_type == "DY":
+                bk = _DY(rk_new, rk_old, pk, dim)
+            else:
+                bk = _BAN(rk_new, rk_old, dim)
+            bk = bk.reshape(shape)
+
+            pk = rk_new + complex_multiplication(bk, pk)
+
+            rk_norm_sq_old = rk_norm_sq_new.clone()
+            rk_old = rk_new.clone()
+
+        return x
+
+    def forward(
+        self,
+        masked_kspace: torch.Tensor,
+        sensitivity_map: torch.Tensor,
+        sampling_mask: torch.Tensor,
+        z: torch.Tensor,
+        lambd: torch.Tensor,
+    ) -> torch.Tensor:
+        """Performs forward pass of :class:`ConjGrad`.
+
+        Parameters
+        ----------
+        masked_kspace : torch.Tensor
+            Masked k-space of shape (N, coil, height, width, complex=2).
+        sensitivity_map : torch.Tensor
+            Coil sensitivities of shape (N, coil, height, width, complex=2).
+        sampling_mask : torch.Tensor
+            Sampling mask of shape (N, 1, height, width, 1).
+        z : torch.Tensor
+            Prediction of image of shape (N, height, width, complex=2).
+        lambd : torch.Tensor
+            Regularaziation (trainable or not) parameter of shape (1).
+
+        Returns
+        -------
+        torch.Tensor
+        """
+        return self.cg(z, masked_kspace, sensitivity_map, sampling_mask, lambd, z)
+
+
+def _PRP(rk_new: torch.Tensor, rk_old: torch.Tensor, dim: List[int]) -> torch.Tensor:
+    r"""Polak-Ribiere-Polyak (PRP) update method for :math:`b_k`:
+
+    .. math ::
+
+        b_k = \frac{ r_{k+1}^{*}(r_{k+1} - r_k) }{ ||r_k||_2^2 } ,
+
+    Parameters
+    ----------
+    rk_new : torch.Tensor
+        Input for :math:`r_{k+1}`.
+    rk_old : torch.Tensor
+        Input for :math:`r_k`.
+    dim : List[int]
+        Dimensions which will be suppressed. Useful when inputs are batched.
+
+    Returns
+    -------
+    bk : torch.Tensor
+        PRP computation for :math:`b_k`.
+    """
+    yk = rk_new - rk_old
+    return complex_division(complex_dot_product(rk_new, yk, dim), complex_dot_product(rk_old, rk_old, dim))
+
+
+def _DY(rk_new: torch.Tensor, rk_old: torch.Tensor, pk: torch.Tensor, dim: List[int]) -> torch.Tensor:
+    r"""Dai-Yuan (DY) update method for :math:`b_k`:
+
+    .. math ::
+
+        b_k = \frac{ ||r_{k+1}||_2^2 }{ p_{k}^{*} y_k } , y_k = r_{k+1} - r_k.
+
+    Parameters
+    ----------
+    rk_new : torch.Tensor
+        Input for :math:`r_{k+1}`.
+    rk_old : torch.Tensor
+        Input for :math:`r_k`.
+    pk : torch.Tensor
+        Input fot :math:`p_k`.
+    dim : List[int]
+        Dimensions which will be suppressed. Useful when inputs are batched.
+
+    Returns
+    -------
+    bk : torch.Tensor
+        DY computation for :math:`b_k`.
+    """
+    yk = rk_new - rk_old
+    return complex_division(complex_dot_product(rk_new, rk_new, dim), complex_dot_product(pk, yk, dim))
```

Comment:
You could assume these inputs are complex valued. You can call `view_as_real` subsequently on it.

Reply:
When we phase to pure complex tensors will do. PLease create an issue and could do it in future commits.